File: services/nlp/app/services/content/gdl.py
"""
Global Digital Library (GDL) content discovery.

Python port of the PHP ``GdlClient``
(``src/Shared/Infrastructure/Http/GdlClient.php``). Searches the GDL content API
(``content.digitallibrary.io``, a WordPress REST API) for openly-licensed
early-grade readers. GDL aggregates content from StoryWeaver, African Storybook
and others under CC-BY / CC-BY-SA. Books are ePUB; this layer only deals in
catalog metadata and the raw ePUB generator URL.

Field mapping is best-effort against the live GDL WordPress API shape: the
captured response lists posts under ``books`` with a ``meta.count`` total, the
reading level surfaces as a ``topic[]`` term named "Level N" (the ``level``
taxonomy is unpopulated), and ``thumbnail`` is ``false`` when a book has no
cover. We look up the post list defensively across a few candidate keys and use
``.get()`` with defaults throughout so missing fields degrade gracefully.
"""

from __future__ import annotations

import html
import re

from app.config import settings
from app.services.http.safe_fetch import safe_get_json

PAGE_SIZE = 20

# Reading-level label, e.g. "Level 3" (case-insensitive, tolerant of spacing).
_LEVEL_PATTERN = re.compile(r"^Level\s*\d+", re.IGNORECASE)
_LEVEL_NUMBER = re.compile(r"(\d+)")


def _decode(text: object) -> str:
    """HTML-entity-decode a GDL text field (titles/descriptions are encoded)."""
    if not text:
        return ""
    return html.unescape(str(text))


def level_to_tier(level: str) -> str:
    """Map a GDL reading level label to a coarse difficulty tier.

    Levels 1-2 -> "easy", 3 -> "medium", 4-5 -> "hard". Returns "" when the
    label carries no number (the PHP client defaults to "medium"; here an
    unknown level yields an empty tier since the result mapping starts from "").
    """
    match = _LEVEL_NUMBER.search(level)
    if not match:
        return ""
    n = int(match.group(1))
    if n <= 2:
        return "easy"
    if n >= 4:
        return "hard"
    return "medium"


def _extract_level(book: dict) -> str:
    """Read the "Level N" label from a book's ``topic[]`` terms ("" if absent)."""
    for topic in book.get("topic") or []:
        if isinstance(topic, dict):
            name = str(topic.get("name", ""))
            if _LEVEL_PATTERN.match(name):
                return name
    return ""


def _first_term_slug(terms: object) -> str:
    """Slug of the first term in a GDL taxonomy array ("" when empty)."""
    if isinstance(terms, list) and terms and isinstance(terms[0], dict):
        return str(terms[0].get("slug", ""))
    return ""


def _first_term_name(terms: object) -> str:
    """Name of the first term in a GDL taxonomy array ("" when empty)."""
    if isinstance(terms, list) and terms and isinstance(terms[0], dict):
        return str(terms[0].get("name", ""))
    return ""


def _thumbnail_url(book: dict) -> str:
    """Cover thumbnail URL; GDL returns ``false`` (not a string) when absent."""
    thumb = book.get("thumbnail")
    return thumb if isinstance(thumb, str) else ""


def _posts_from_response(response: dict) -> list:
    """Extract the post list defensively across candidate GDL response keys."""
    for key in ("hits", "results", "books"):
        value = response.get(key)
        if isinstance(value, list):
            return value
    return []


def _map_post(book: dict) -> dict:
    """Normalise a raw GDL post record into the API result shape."""
    level = _extract_level(book)
    return {
        "id": int(book.get("postId") or 0),
        "title": _decode(book.get("title", "")),
        "publisher": _decode(book.get("publisher", "")),
        "description": _decode(book.get("description", "")).strip(),
        "language": _first_term_slug(book.get("language")),
        "license": _first_term_name(book.get("license")),
        "level": level,
        "difficultyTier": level_to_tier(level),
        "thumbnail": _thumbnail_url(book),
        "sourceUri": str(book.get("postLink") or ""),
        "epubUrl": str(book.get("epubUrl") or "").strip(),
    }


async def search(
    query: str, language_code: str | None = None, page: int = 1
) -> dict:
    """Search the Global Digital Library catalog.

    Top-level keys: ``results`` (list), ``count`` (total across pages, from
    ``meta.count``), ``next`` (whether ``page * PAGE_SIZE`` < ``count``).
    """
    params: dict[str, object] = {}
    if query:
        params["query"] = query
    if language_code:
        params["language"] = language_code.lower()
    if page > 1:
        params["_skip"] = (page - 1) * PAGE_SIZE

    response = await safe_get_json(settings.gdl_base_url, params=params)
    if not isinstance(response, dict):
        response = {}

    posts = _posts_from_response(response)
    results = [_map_post(p) for p in posts if isinstance(p, dict)]

    meta = response.get("meta")
    if isinstance(meta, dict):
        count = int(meta.get("count") or len(results))
    else:
        count = len(results)

    return {
        "results": results,
        "count": count,
        "next": (page * PAGE_SIZE) < count,
    }


async def browse(language_code: str, page: int = 1) -> dict:
    """Browse books for a language (search with an empty query)."""
    return await search("", language_code, page)

File: src/frontend/js/shared/offline/local/content/gdl.ts
/**
 * Global Digital Library (GDL) catalog client, ported to run client-side.
 *
 * Faithful port of `services/nlp/.../content/gdl.py` (itself a port of the PHP
 * `GdlClient`): searches the GDL WordPress content API for openly-licensed
 * early-grade readers (StoryWeaver, African Storybook, …), reading the "Level N"
 * label from each book's `topic[]` terms and mapping it to a difficulty tier.
 *
 * Browse/suggest only: this client supplies the catalog row. The EPUB books it
 * links to are imported on-device by the sibling `epub.ts` module (download →
 * unzip → extract → parse), wired through the content repository's
 * `importEpubText` (see both repos' BRIEFING.md seam).
 *
 * @license Unlicense <http://unlicense.org/>
 */

import { corsFreeGetJson } from './native-http';
import type { CatalogPage } from './gutenberg';
import type { DifficultyTier } from './difficulty';

const GDL_URL = 'https://content.digitallibrary.io/wp-json/content-api/v1/contentsearch';
const PAGE_SIZE = 20;

/** A GDL book as the suggestion UI consumes it. */
export interface GdlBook {
  id: number;
  title: string;
  publisher: string;
  description: string;
  language: string;
  license: string;
  level: string;
  difficultyTier: DifficultyTier | '';
  thumbnail: string;
  sourceUri: string;
  epubUrl: string;
}

interface GdlTerm {
  name?: string;
  slug?: string;
}

interface GdlPost {
  postId?: number;
  title?: string;
  publisher?: string;
  description?: string;
  language?: GdlTerm[];
  license?: GdlTerm[];
  topic?: GdlTerm[];
  thumbnail?: string | boolean;
  postLink?: string;
  epubUrl?: string;
}

interface GdlResponse {
  hits?: GdlPost[];
  results?: GdlPost[];
  books?: GdlPost[];
  meta?: { count?: number };
}

// Reading-level label, e.g. "Level 3" (case-insensitive, tolerant of spacing).
const LEVEL_PATTERN = /^Level\s*\d+/i;
const LEVEL_NUMBER = /(\d+)/;

/**
 * Map a GDL reading-level label to a coarse tier: 1-2 -> easy, 3 -> medium,
 * 4-5 -> hard. Returns '' when the label carries no number. Mirrors
 * `gdl.py::level_to_tier`.
 */
export function levelToTier(level: string): DifficultyTier | '' {
  const match = LEVEL_NUMBER.exec(level);
  if (!match) {
    return '';
  }
  const n = parseInt(match[1], 10);
  if (n <= 2) {
    return 'easy';
  }
  if (n >= 4) {
    return 'hard';
  }
  return 'medium';
}

/** Read the "Level N" label from a book's `topic[]` terms ('' if absent). */
function extractLevel(book: GdlPost): string {
  for (const topic of book.topic ?? []) {
    const name = String(topic?.name ?? '');
    if (LEVEL_PATTERN.test(name)) {
      return name;
    }
  }
  return '';
}

let decoderEl: HTMLTextAreaElement | null = null;

/** HTML-entity-decode a GDL text field (titles/descriptions are encoded). */
function decode(text: unknown): string {
  if (!text) {
    return '';
  }
  const raw = String(text);
  if (typeof document === 'undefined') {
    return raw;
  }
  if (!decoderEl) {
    decoderEl = document.createElement('textarea');
  }
  decoderEl.innerHTML = raw;
  return decoderEl.value;
}

function firstTermSlug(terms: GdlTerm[] | undefined): string {
  return terms && terms.length > 0 ? String(terms[0]?.slug ?? '') : '';
}

function firstTermName(terms: GdlTerm[] | undefined): string {
  return terms && terms.length > 0 ? String(terms[0]?.name ?? '') : '';
}

function mapPost(book: GdlPost): GdlBook {
  const level = extractLevel(book);
  return {
    id: Number(book.postId) || 0,
    title: decode(book.title),
    publisher: decode(book.publisher),
    description: decode(book.description).trim(),
    language: firstTermSlug(book.language),
    license: firstTermName(book.license),
    level,
    difficultyTier: levelToTier(level),
    thumbnail: typeof book.thumbnail === 'string' ? book.thumbnail : '',
    sourceUri: String(book.postLink ?? ''),
    epubUrl: String(book.epubUrl ?? '').trim(),
  };
}

function postsFromResponse(response: GdlResponse): GdlPost[] {
  return response.hits ?? response.results ?? response.books ?? [];
}

/**
 * Search (or, with an empty query, browse) the Global Digital Library. `page`
 * is 1-based; GDL paginates via a 0-based `_skip` offset of `PAGE_SIZE`.
 */
export async function searchGdl(
  query: string,
  languageCode: string | null,
  page: number
): Promise<CatalogPage<GdlBook>> {
  const params = new URLSearchParams();
  if (query) {
    params.set('query', query);
  }
  if (languageCode) {
    params.set('language', languageCode.toLowerCase());
  }
  if (page > 1) {
    params.set('_skip', String((page - 1) * PAGE_SIZE));
  }
  const qs = params.toString();
  const response = await corsFreeGetJson<GdlResponse>(GDL_URL + (qs ? `?${qs}` : ''));

  const posts = postsFromResponse(response).filter((p) => p && typeof p === 'object');
  const results = posts.map(mapPost);
  const count = response.meta?.count ?? results.length;

  return {
    results,
    count,
    next: page * PAGE_SIZE < count,
  };
}
